polygeist/lib/Dialect/Polygeist/Analysis/SYCLIDAndRangeAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+
+namespace mlir {
+namespace polygeist {
+
+//===----------------------------------------------------------------------===//
+// IDRangeInformation
+//===----------------------------------------------------------------------===//
+
+raw_ostream &operator<<(raw_ostream &os, const IDRangeInformation &info) {
+  if (!info.hasFixedDimensions()) {
+    os.indent(4) << "<unknown>\n";
+    return os;
+  }
+
+  if (info.isConstant()) {
+    os.indent(4) << "constant<";
+    llvm::interleaveComma(info.getConstantValues(), os);
+    os << ">\n";
+    return os;
+  }
+
+  os.indent(4) << "fixed<" << info.getNumDimensions() << ">\n";
+  return os;
+}
+
+IDRangeInformation::IDRangeInformation()
+    : dimensions{std::nullopt}, constantValues{std::nullopt} {}
+
+IDRangeInformation::IDRangeInformation(size_t dim)
+    : dimensions{dim}, constantValues{std::nullopt} {}
+
+IDRangeInformation::IDRangeInformation(llvm::ArrayRef<size_t> constVals)
+    : dimensions{constVals.size()}, constantValues{constVals} {}
+
+bool IDRangeInformation::hasFixedDimensions() const {
+  return dimensions.has_value();
+}
+
+size_t IDRangeInformation::getNumDimensions() const {
+  assert(dimensions.has_value() &&
+         "Requesting fixed dimensions from non-fixed id/range");
+  return *dimensions;
+}
+
+bool IDRangeInformation::isConstant() const {
+  return constantValues.has_value();
+}
+
+const llvm::SmallVector<size_t, 3> &
+IDRangeInformation::getConstantValues() const {
+  assert(constantValues.has_value() &&
+         "Requesting constant values from non-constant id/range");
+  return *constantValues;
+}
+
+const IDRangeInformation
+IDRangeInformation::join(const IDRangeInformation &other) const {
+  if (isConstant() && other.isConstant()) {
+    if (getConstantValues() == other.getConstantValues())
+      return *this;
+  }
+  if (hasFixedDimensions() && other.hasFixedDimensions()) {
+    if (getNumDimensions() == other.getNumDimensions())
+      return IDRangeInformation{getNumDimensions()};
+  }
+  return IDRangeInformation{};
+}
+
+//===----------------------------------------------------------------------===//
+// Helpers
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+static std::optional<int> getConstantUInt(Value v) {
+  Operation *op = v.getDefiningOp();
+  if (!op)
+    return std::nullopt;
+
+  if (!op->hasTrait<OpTrait::ConstantLike>())
+    return std::nullopt;
+
+  llvm::SmallVector<OpFoldResult> folded;
+  if (failed(op->fold({}, folded)) || folded.size() != 1)
+    return std::nullopt;
+
+  if (!folded.front().is<Attribute>() ||
+      !isa<IntegerAttr>(folded.front().get<Attribute>()))
+    return std::nullopt;
+
+  return cast<IntegerAttr>(folded.front().get<Attribute>()).getInt();
+}
+
+template <typename Type> bool isConstructor(const Definition &def) {
+  if (!def.isOperation())
+    return false;
+
+  auto constructor = dyn_cast<sycl::SYCLHostConstructorOp>(def.getOperation());
```

Comment:
The analysis is limited to host constructor operation, can we extend it to make it work also on SYCLConstructorOp ?

Reply:
We can probably extend it, it should probably work. 

Would this be something that is useful for you ad-hoc? If you do not have an immediate use-case, we could add that support in a separate PR. If you want to have it now, I can look into adding it here, including tests.

Reply:
No immediate use-case. Is OK to extend it in a later PR.